types/index.d.ts

```diff
@@ -0,0 +1,94 @@
+// TypeScript Version: 3.5
+
+import {Node, Parent} from 'unist'
+
+interface Visitor<V extends Node> {
+  (node: V, index: number, parent: Node): void
+}
+
+/**
+ * Check that type property matches expectation for a node
+ *
+ * @typeParam T type of node that passes test
+ */
+type TestType<T extends Node> = T['type']
+
+/**
+ * Check that some attributes on a node are matched
+ *
+ * @typeParam T type of node that passes test
+ */
+type TestObject<T extends Node> = Partial<T>
+
+/**
+ * Check if a node passes a test
+ *
+ * @typeParam T type of node that passes test
+ */
+interface TestFunction<T extends Node> {
+  /**
+   * Check if a node passes a test
+   *
+   * @param node node to check
+   * @param index index of node in parent
+   * @param parent parent of node
+   */
+  (node: Node, index?: number, parent?: Parent): node is T
+}
+
+/**
+ * Union of all the types of tests
+ *
+ * @typeParam T type of node that passes test
+ */
+type Test<T extends Node> = TestType<T> | TestObject<T> | TestFunction<T>
```

Comment:
These three types similar to and partially duplicate https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0bb210867d16170c4a08d9ce5d132817651a0f80/types/unist-util-is/index.d.ts#L11-L19
With the addition of generics and some additional type narrowing

And will be needed again in:
* https://github.com/syntax-tree/unist-util-visit-parents
* https://github.com/syntax-tree/unist-util-remove
* https://github.com/syntax-tree/unist-util-filter
* https://github.com/syntax-tree/unist-util-find-all-before
* https://github.com/syntax-tree/unist-util-find-all-after
* https://github.com/syntax-tree/unist-util-visit-children
* https://github.com/syntax-tree/unist-util-find-before
* https://github.com/syntax-tree/unist-util-find-after
* and more

There are some options for this:
* add helper methods to `unist` as utils
* upstream to `unist-util-is` and reuse
* create a new `unist-help-functions` repo
* leave them duplicated

thoughts? :thought_balloon:

Reply:
I'm leaning toward upstreaming to `unist-util-is`

Reply:
This could be refactored out if/when https://github.com/syntax-tree/unist-util-is/pull/9 is merged

Reply:
@wooorm @ChristianMurphy About the above module list, I think it would be nice if we have a project board in unified organization so we can track progress of defining types for each module.

Reply:
> This could be refactored out if/when syntax-tree/unist-util-is#9 is merged

Sounds good to me. I think we should wait until syntax-tree/unist-util-is#9. There are still some issues to be discussed.

Reply:
@Rokt33r would an issue with a checklist of all the modules work as well? E.g., in `syntax-tree/ideas`? (and later in other orgs, like `vfile/ideas`?)

Reply:
Having a tracker issue with a checklist sounds like a good idea. :+1: 
It would offer a place to link folks who are curious ([for example this Spectrum thread](https://spectrum.chat/unified/type-definitions/how-to-get-hast-element-type-definition~f16a9774-f826-410a-9e2b-d7dd0701f9bd)) and allow for more easily coordinating.

Reply:
> I'm leaning toward upstreaming to `unist-util-is`

Types have been added into `unist-util-is` once the next release is out, this PR can be updated to leverage the utility types from upstream.

Reply:
> would an issue with a checklist of all the modules work as well? E.g., in syntax-tree/ideas? (and later in other orgs, like vfile/ideas?)

@wooorm I guess we should make one in unified organization because it would cover whole unified ecosystem. So unified/ideas might be enough?

Reply:
The reason I’d suggest per-organisation issues/lists is that they’re a bit smaller. There are at least 250 repos under the whole collective, so one issue for all of them would, I think, become too unwieldy, with many comments, and take a year to close.

But they both work, so feel free to create one!

Reply:
`unist-util-is@4.0.0` released!